src/messages/workspace_symbol.cc

```diff
@@ -51,12 +57,115 @@ struct Out_WorkspaceSymbol : public lsOutMessage<Out_WorkspaceSymbol> {
+    return kCamelScore;
+  if (curr == CharClass::NonWord)
+    return kNonWordScore;
+  return 0;
+}
+
+/*
+fuzzyEvaluate implements a global sequence alignment algorithm to find the maximum accumulated score by aligning `pattern` to `str`. It applies when `pattern` is a subsequence of `str`.
+
+Scoring criteria
+- Prefer matches at the start of a word, or the start of subwords in CamelCase/camelCase/camel123 words. See kWordStartScore/kCamelScore
+- Non-word characters matter. See kNonWordScore
+- The first characters of words of `pattern` receive bonus because they usually have more significance than the rest. See kPatternStartMultiplier
+- Superfluous characters in `str` will reduce the score (gap penalty). See kGapScore
+- Prefer early occurrence of the first character. See kLeadingGapScore/kGapScore
+
+The recurrence of the dynamic programming:
+dp[i][j]: maximum accumulated score by aligning pattern[0..i] to str[0..j]
+dp[0][j] = leading_gap_penalty(0, j) + score[j]
+dp[i][j] = max(dp[i-1][j-1] + CONSECUTIVE_SCORE, max(dp[i-1][k] + gap_penalty(k+1, j) + score[j] : k < j))
+The first dimension can be suppressed since we do not need a matching scheme, which reduces the space complexity from O(N*M) to O(M)
+*/
+int fuzzyEvaluate(const std::string& pattern,
+                  const std::string& str,
+                  std::vector<int>& score,
+                  std::vector<int>& dp) {
+  bool pstart = true;  // whether we are aligning the start of a word in pattern
+  int uleft = 0,       // value of the upper left cell
+      ulefts = 0,      // maximum value of uleft and cells on the left
+      left, lefts;     // similar to uleft/ulefts, but for the next row
+
+  // Calculate position score for each character in str.
+  CharClass prev = CharClass::NonWord;
+  for (int i = 0; i < int(str.size()); i++) {
+    CharClass cur = getCharClass(str[i]);
+    score[i] = getScoreFor(prev, cur);
+    prev = cur;
+  }
+  std::fill_n(dp.begin(), str.size(), kMinScore);
+
+  // Align each character of pattern.
+  for (unsigned char pc: pattern) {
+    if (isspace(pc)) {
+      pstart = true;
+      continue;
+    }
+    lefts = kMinScore;
+    // Enumerate the character in str to be aligned with pc.
+    for (int i = 0; i < int(str.size()); i++) {
+      left = dp[i];
+      lefts = std::max(lefts + kGapScore, left);
+      if (tolower(pc) == tolower(str[i])) {
+        int t = score[i] * (pstart ? kPatternStartMultiplier : 1);
+        dp[i] = i == 0 ? kLeadingGapScore * i + t
+                       : std::max(uleft + kConsecutiveScore, ulefts + t);
+      } else
+        dp[i] = kMinScore;
+      uleft = left;
+      ulefts = lefts;
+    }
+    pstart = false;
+  }
+
+  // Enumerate the end position of the match in str, incurring kGapScore for
+  // discarded characters.
+  lefts = kMinScore;
+  for (int i = 0; i < int(str.size()); i++)
+    lefts = std::max(lefts + kGapScore, dp[i]);
+  return lefts;
+}
+
```

Comment:
I think these functions should belong to a separate file (e.g. utils.cc strings.cc fuzzy.cc). And which algorithm is this function using? I am not familiar with fuzzy matching, but what is the one used in `fzf` or some alternatives?

Reply:
fzf's https://github.com/junegunn/fzf/blob/master/src/algo/algo.go is too complicated... I think this simple sequence alignment algorithm should be good enough.